### ISSUE TITLE
Ignore "Replace unneeded comprehension with generator" when used with `str.join()`

### ISSUE THREAD

**Issue by Bluenix2** [sourcery-ai/sourcery#150]
### Issue description or question

Currently Sourcery is recommending turning the following code, that is using a list comprehension-
```py
include_roles = ', '.join([str(int(r)) for r in roles]) if isinstance(roles, list) else roles
```
Into the much slower version using a generator
```py
include_roles = (
    ', '.join(str(int(r)) for r in roles)
    if isinstance(roles, list)
    else roles
)
```
*I'll format it back so that they're comparable*
```py
include_roles = ', '.join(str(int(r)) for r in roles) if isinstance(roles, list) else roles
```

This is actually slower, which seems weird for Sourcery to recommend. For that reason, while the error shouldn't be removed, a manual gate disabling this could be added.

I am not sure how Sourcery is implemented, this may be a trivial thing or very complicated.

### Sourcery Version

![image](https://user-images.githubusercontent.com/38372706/130213114-4ab77241-0464-4403-867d-ec3e9cb15bbb.png)

### OS name and version

Windows

**Comment by Hellebore** [comment]
Hi - thanks for raising!

We're primarily concerned with readability but can take performance into account. Have you got some up-to-date evidence that it's slower? Had a quick look and the stackoverflow stuff at least was for quite old Python versions.

**Comment by Bluenix2** [comment]
```py
C:\Users\%username%>python -m timeit -n 100 "', '.join(str(i) for i in range(100_000))"
100 loops, best of 5: 26.1 msec per loop

C:\Users\%username%>python -m timeit -n 100 "', '.join([str(i) for i in range(100_000)])"
100 loops, best of 5: 22.2 msec per loop
```
The most basic example I could make. The difference can be greater depending on what you do, and with close to no readability hit I think this could be preferred.

This is Python 3.7 on my machine.

**Comment by ruancomelli** [comment]
This has a lot to do with the implementation of `', '.join(...)`. In fact, some functions have a much better performance when called with sequences (such as `list`s and `tuple`s) when compared with plain generators or iterables with unknown length.

To give you another example, suppose that we want a function that calculates the length of any iterable:
```py
from typing import Iterable

def length(iterable: Iterable) -> int:
    try:
        # shortcut when `iterable` has a `__len__`
        # `list`s, `tuple`s and `range`s enter here
        return len(iterable)
    except TypeError:
        # `iterable` has no `__len__`, so we have to count
        # its elements by hand
        # this is the case for generators and iterables in general
        for length_candidate, _ in enumerate(iterable, start=1):
            pass
        return length_candidate
```

Comparing execution time:
```py
>>> import timeit

>>> print("length([x for x in range(100_000)]):", timeit.timeit("length([x for x in range(100_000)])", number=100, globals={'length': length}))
length([x for x in range(100_000)]): 0.43636004500058334
>>> print("length(x for x in range(100_000)):", timeit.timeit("length(x for x in range(1_000_000))", number=100, globals={'length': length}))
length(x for x in range(100_000)): 7.137552369999867
```
Lists comprehensions are 16x faster than a plain generator. Interestingly, Sourcery *does not* suggest replacing `length([x for x in range(100)])` with a generator in my case (any reasons why not?). Nonetheless, I believe that `str.join()` gets some performance benefits when called on sequences, and that would explain the boost in speed.

---
Using Python 3.7 and Sourcery 0.9.4 on Ubuntu 20.04.3 LTS.

**Comment by Hellebore** [comment]
Addressed in 0.9.6.
